tooling/cli/src/profiling_alloc.rs

```diff
@@ -0,0 +1,172 @@
+        self.current.load(Ordering::SeqCst)
+    }
+
+    pub fn max(&self) -> usize {
+        self.max.load(Ordering::SeqCst)
+    }
+
+    pub fn reset_max(&self) -> usize {
+        let current = self.current();
+        self.max.store(current, Ordering::SeqCst);
+        current
+    }
+
+    pub fn count(&self) -> usize {
+        self.count.load(Ordering::SeqCst)
+    }
+
+    #[cfg(feature = "tracy")]
+    pub fn enable_tracy(&self, depth: usize) {
+        self.tracy_enabled.store(true, Ordering::SeqCst);
+        self.tracy_depth.store(depth, Ordering::SeqCst);
+    }
+
+    #[cfg(feature = "tracy")]
+    pub fn disable_tracy(&self) {
+        self.tracy_enabled.store(false, Ordering::SeqCst);
+    }
+
+    #[allow(unused_variables)] // Conditional compilation may not use all variables
+    fn tracy_alloc(&self, size: usize, ptr: *mut u8) {
+        // If Tracy profiling is enabled, report this allocation to Tracy.
+        #[cfg(feature = "tracy")]
+        if self.tracy_enabled.load(Ordering::SeqCst) {
+            let depth = self.tracy_depth.load(Ordering::SeqCst);
+            if depth == 0 {
+                // If depth is 0, we don't capture any stack information
+                unsafe {
+                    tracy_sys::___tracy_emit_memory_alloc(ptr.cast(), size, 1);
+                }
+            } else {
+                // Capture stack information up to `depth` frames
+                unsafe {
+                    tracy_sys::___tracy_emit_memory_alloc_callstack(
+                        ptr.cast(),
+                        size,
+                        depth as i32,
+                        1,
+                    );
+                }
+            }
+        }
+    }
+
+    #[allow(unused_variables)] // Conditional compilation may not use all variables
+    fn tracy_dealloc(&self, ptr: *mut u8) {
+        // If Tracy profiling is enabled, report this deallocation to Tracy.
+        #[cfg(feature = "tracy")]
+        if self.tracy_enabled.load(Ordering::SeqCst) {
+            let depth = self.tracy_depth.load(Ordering::SeqCst);
+            if depth == 0 {
+                // If depth is 0, we don't capture any stack information
+                unsafe {
+                    tracy_sys::___tracy_emit_memory_free(ptr.cast(), 1);
+                }
+            } else {
+                // Capture stack information up to `depth` frames
+                unsafe {
+                    tracy_sys::___tracy_emit_memory_free_callstack(ptr.cast(), depth as i32, 1);
+                }
+            }
+        }
+    }
+}
+
+#[allow(unsafe_code)]
+unsafe impl GlobalAlloc for ProfilingAllocator {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let ptr = SystemAlloc.alloc(layout);
+        let size = layout.size();
+        let current = self
+            .current
+            .fetch_add(size, Ordering::SeqCst)
+            .wrapping_add(size);
+        self.max.fetch_max(current, Ordering::SeqCst);
```

Comment:
Nit, between `fetch_add` and `fetch_max`, another thread could update `max`, causing the local current calculation to be stale, we can avoid this by using `fetch_update`

```suggestion
        self.max.fetch_update(Ordering::SeqCst, Ordering::SeqCst, |max_val| {
    let new_current = self.current.load(Ordering::SeqCst);
    (new_current > max_val).then_some(new_current)
}).ok();
```